R/cmd_methods.R

```diff
@@ -113,26 +113,69 @@ monocle_create <- function(
         message('You need to provide the expression matrix by --expression-matrix. Aborting.')
         q(save = 'no', status = 1)
     }
-    #the three constituents of the new_cell_data_set() call
-    #are passed as arguments to the function, and live in this list
-    for (var in c('expression_matrix', 'cell_metadata', 'gene_annotation'))
-    {
-        file = createCDS_options[[var]]
-        if (is.null(file))
-            assign(var, NULL)
-        else
-        {
-            if (toupper(substr(file, nchar(file)-2, nchar(file))) == 'RDS')
-                assign(var, readRDS(file))
-            else if (toupper(substr(file, nchar(file)-2, nchar(file))) == 'MTX')
-                assign(var, Matrix::readMM(file))
-            else if (toupper(substr(file, nchar(file)-2, nchar(file))) == 'TSV')
-                assign(var, as.matrix(read.delim(file, row.names = 1, stringsAsFactors = FALSE)))
-            else
-                assign(var, as.matrix(read.csv(file, row.names = 1, stringsAsFactors = FALSE)))
+
+    .parse_input_data = function(file_type){
+        if(! file_type %in% c('expression_matrix', 'cell_metadata', 'gene_annotation'))
+            stop(paste("Incorrect argument provided", file_type))
+
+        file = createCDS_options[[file_type]]
+
+        if (is.null(file)) 
+            return(NULL)
+        ext = toupper(substr(file, nchar(file)-2, nchar(file)))
+        if (ext == 'RDS') 
+            return(readRDS(file))
+        if (ext == 'MTX')
+            return(Matrix::readMM(file))
+
+        # handle different text formats
+        if (ext == 'TSV') 
+            sep = "\t"
+        else sep = ","
+
+        # Read file once, and adjust row/columns as appropriate before
+        # returning as matrix
+
+        if ( file_type == 'expression_matrix' ){
+            mat = as.matrix(read.delim(file, sep = sep, stringsAsFactors = FALSE, row.names = 1)) 
+        }else{
+            mat = read.delim(file, sep = sep, stringsAsFactors = FALSE, row.names = NULL) 
+            dim_to_check = ifelse(file_type == 'cell_metadata', 2, 1)
+
+            expected_rows = dim(expression_matrix)[dim_to_check]
+
+            if (nrows(mat) != expected_rows){
+                if (nrows(mat) - 1 == expected_rows){
+                    mat <- mat[-1,]
+                }else{
+                    message(paste('Number of', file_type, 'entries not appropriate to matrix.'))
+                    q(save = 'no', status = 1)
+                }
+            }
+
+            # Move first column into rownames where possible
```

Comment:
`new_cell_data_set()` seems to be able to handle empty metadata table as long as the row names exist, and `as.matrix()` keeps rownames, so `as.matrix(..., read.delim(row.names=1))` works fine. But the approach here is also ok.

Reply:
Ahh, I hadn't tested that, but was told it wouldn't work - now simplified.